[Merged by Bors] - feat: extensible `push` and `pull` tactics

JovanGerb opened the pull request:
This PR defines the `push` and `pull` tactics, and makes `push_neg` a macro for `push Not`. The tactics are also available in `conv` mode.

For tagging, there is only the `@[push]` attribute, which adds the reverse rewrite for the `pull` tactic when relevant. In the future, we may also need a separate `@[pull X]` attribute for pulling `X`.

Thanks to this change, we will be able to make `push_neg` into a more powerful tactic by tagging more lemmas. It also means that about 60 files now don't need to import `LinearOrder`/`PartialOrder`. This will be especially useful when we get the `@[to_dual]` attribute.

This work originally started in #21769. There is now the follow up PR #29000 which adds `push` tags and tests for them.

The `@[push]` attribute is defined in `Mathlib.Tactic.Push.Attr` and the main implementation of `push` and `pull` is in `Mathlib.Tactic.Push`.

Some proofs need to be fixed because the new simp-based `push_neg` can see through more reducible definitions.

Zulip conversation: [#mathlib4 > I made an extensible &#96;push&#96; tactic generalizing &#96;push_neg&#96;](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/I.20made.20an.20extensible.20.60push.60.20tactic.20generalizing.20.60push_neg.60)

closes #21841

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

JovanGerb committed:
feat: extensible `push`(_neg) tactic - part 1

JovanGerb committed:
`ne_eq` rewrite can be done with just the lemma.

JovanGerb committed:
feat: debug command `#push_discr_tree`

JovanGerb commented:
What should I do with the copyright docstring at the top of these files? I added myself as the main author, set the year to 2025, and left the names which were there in the original file.

grunweg commented:
> What should I do with the copyright docstring at the top of these files? I added myself as the main author, set the year to 2025, and left the names which were there in the original file.

Adding yourself as an author and leaving the original authors sounds good. Please leave the year as it was.

JovanGerb committed:
fix

JovanGerb committed:
.

JovanGerb commented:
!bench

JovanGerb committed:
undo changes to help merge

JovanGerb committed:
Merge remote-tracking branch 'origin' into push_tactic

JovanGerb committed:
redo change to import (after merge)

JovanGerb committed:
.

JovanGerb committed:
.

JovanGerb committed:
Merge remote-tracking branch 'origin' into push_tactic

JovanGerb committed:
.

grunweg commented on Mathlib/Tactic/Push.lean:
```diff
@@ -1,237 +1,301 @@
 
-Note that names are conserved by this tactic, contrary to what would happen with `simp`
-using the relevant lemmas.
+/-- Retreave the `Head` of an expression. -/
```
```suggestion
/-- Retrieve the `Head` of an expression. -/

JovanGerb committed:
typo

Co-authored-by: grunweg <rothgami@math.hu-berlin.de>

JovanGerb committed:
Merge branch 'master' into push_tactic

JovanGerb committed:
add missing import

JovanGerb committed:
Merge branch 'master' into push_tactic

JovanGerb committed:
Merge branch 'master' into push_tactic

JovanGerb committed:
fix proofs

JovanGerb committed:
.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

JovanGerb committed:
Merge branch 'master' into push_tactic

j-loreaux reviewed (commented):
Please merge master to fix the conflicts.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

JovanGerb committed:
incorporate some suggestions

JovanGerb committed:
Merge branch 'master' into push_tactic

JovanGerb commented on Mathlib/Tactic/Push.lean:
```diff
@@ -1,237 +1,301 @@
+  withTheReader Term.Context ({ · with ignoreTCFailures := true }) <|
+  Term.withoutModifyingElabMetaStateWithInfo <|
+  Term.withoutErrToSorry <|
+  Term.elabTerm stx none
```
The term passed to `push` is only there to determine the head constant. But I will refactor this to be the same as in `simp`, where it is allowed to write something like `simp only [· + ·]`.

JovanGerb committed:
fixes

JovanGerb committed:
more cleanup

JovanGerb committed:
new syntax for specifying the `Head`

JovanGerb committed:
feat: `#push`

JovanGerb committed:
fix

JovanGerb committed:
feat: `pull` environment extension

JovanGerb committed:
feat: `pull` tactic

JovanGerb committed:
remove `pushNeg` simproc as it is redundant

JovanGerb committed:
add test for `pull`

JovanGerb committed:
cleanup

JovanGerb commented:
@j-loreaux, thank you for your review. I think I have addressed all of you comments.

I have also further refactored the code:
- I defined the `pull` tactic as a converse to the `push` tactic.
- I changed the syntax of the `push` argument. Now instead of `_ ∈ _` and `∀ _, _` it should be `· ∈ ·` and `∀ _, ·`. This is consistent with the `simp [· ∈ ·]` syntax.
- I removed the `pushNeg` simproc because all `push_neg` lemmas are already tagged with `@[simp]` anyways.

JovanGerb commented:
There is some mysterious CI failure...

JovanGerb committed:
Merge branch 'master' into push_tactic

j-loreaux reviewed (commented):
This is very nice Jovan!

JovanGerb commented on Mathlib/Tactic/Push.lean:
```diff
@@ -1,238 +1,422 @@
+
+In addition to constants, `push` can be used to push `∀` and `fun` binders:
+- `push ∀ _, ·` rewrites `∀ a, p a ∧ q a` into `(∀ a, p a) ∧ (∀ a, q a)`.
+- `push fun _ ↦ ·` rewrites  `fun x => f x + g x` into `f + g`
```
So, that would be a `pushFun` simproc?

JovanGerb replied:
Note that the current implementation of `push fun _ ↦ ·`/`simp [pushFun]` is very inefficient: it will try every `fun` lemma on every `fun` in the expression.

I hadn't thought of doing this before, but this issue can be fixed by switching to `RefinedDiscrTree`.

JovanGerb commented on Mathlib/Tactic/Push.lean:
```diff
@@ -1,238 +1,422 @@
+`push` pushes the given constant away from the root of the expression. For example
+- `push · ∈ ·` rewrites `x ∈ {y} ∪ zᶜ` into `x = y ∨ ¬ x ∈ z`.
+- `push (disch := positivity) Real.log` rewrites `log (a * b ^ 2)` into `log a + 2 * log b`.
+- `push ¬·` is the same as `push_neg` or `push Not`, and it rewrites
```
I never know whether to put a space after `¬`. I'll add it in.

plp127 replied:
The pretty printer never puts a space after `¬`

JovanGerb commented on Mathlib/Tactic/Push.lean:
```diff
@@ -1,238 +1,422 @@
+def pushSimpConfig : Simp.Config where
+  zeta := false
+  proj := false
+  congrConsts := false -- this is a workaround, and can hopefully be removed
```
I'll remove it and see if anything breaks

JovanGerb committed:
remove workaround

JovanGerb committed:
mention special casing of `Not`

JovanGerb committed:
fix spacing

JovanGerb commented on Mathlib/Tactic/Push.lean:
```diff
@@ -1,238 +1,422 @@
+
+/--
+`push` pushes the given constant away from the root of the expression. For example
+- `push · ∈ ·` rewrites `x ∈ {y} ∪ zᶜ` into `x = y ∨ ¬ x ∈ z`.
```
I could modify the syntax to require brackets around this notation. Either (· ∈ ·) or [· ∈ ·] (which matches simp [· ∈ ·]). But they aren't necessary. If we go for square brackets, we could also allow passing multiple arguments to `push`, but I don't really know how that would be used.

JovanGerb commented on MathlibTest/push.lean:
```diff
@@ -0,0 +1,71 @@
+  guard_target =ₐ p ∨ ∃ n, q ∧ n = 1
+  exact test_sorry
+
+-- the following examples still need more tagging to work
+
```
Would it be easier to review if more @[push] tags and accompanying tests were added in this PR, or in a follow-up PR?

j-loreaux replied:
A follow-up PR is probably better, but you can link to it from the PR description here so that reviewers know to look for it.

JovanGerb committed:
add `pushFun` simproc, and move some code

JovanGerb committed:
tag lemmas for `pushFun` and add test

JovanGerb committed:
add tags and test for `push · ∈ ·`

JovanGerb committed:
fix

JovanGerb commented:
I've now tagged some lemmas for `push · ∈ ·` and `push fun _ ↦ ·` and added tests for them. I've also added the `pushFun` simproc (should it have a different name? `normFun`?).
- I should switch to `RefinedDiscrTree` so that `pushFun` becomes more efficient.
- In the current implementation, `pull fun _ ↦ ·` doesn't work. But I think that one is less usefull than `push fun _ ↦ ·`.
- Should I add a way to limit the number of rewrites done by `push`? That could be useful e.g. for `push_neg`. Then we should come up with a syntax for this. This can also be a later PR.
- The syntax `push · ∈ ·` looks a bit funny without brackets around the dots. We could still decide to modify it to `push [· ∈ ·]`. And then `push [· ∈ ·, Not]` could for example mean to push membership and negation at the same time. But allowing this is also dangerous because certain combinations of constants to push will cause infinite loops.
- Further constants to be tagged would include `Real.log`, `Real.exp`, `⁻¹`, `Finset.sum`/`Finset.prod`, `tsum`/`tprod`, and extending tagging for `· ∈ ·` from `Set` to `Finset` (and to other containers)

JovanGerb committed:
Merge branch 'master' into push_tactic

JovanGerb committed:
switch to using `transformAtLocation`

JovanGerb committed:
fix proof

JovanGerb committed:
fix test after switching to `transformAtLocation`

JovanGerb committed:
merge if-condition

JovanGerb committed:
fix elaboration of `(disch := ...)`

JovanGerb committed:
add tests for log

JovanGerb committed:
also push `mem_image` and friends

JovanGerb committed:
Merge branch 'master' into push_tactic

JovanGerb committed:
.

JovanGerb committed:
fix

JovanGerb committed:
fix `pushFun`

JovanGerb committed:
use `push only`

JovanGerb committed:
fix

JovanGerb committed:
.

JovanGerb committed:
revert extra tags and tests

JovanGerb committed:
restore necessary tags

JovanGerb committed:
fix `pull fun _ => ·`

JovanGerb committed:
clean up

JovanGerb committed:
fix style and spelling

JovanGerb committed:
new elaborator for `push` argument: enforce use of `_`

JovanGerb committed:
elaborate the argument, to get better error messages

JovanGerb committed:
Merge branch 'master' into push_tactic

JovanGerb commented:
Following the Zulip conversation, the syntax is now `push _ ∈ _`. The elaborator that enforces that `_` is used instead of an explicit term is a bit hacky, but I think works well enough.

j-loreaux commented on MathlibTest/push.lean:
```diff
@@ -0,0 +1,49 @@
+  guard_target =ₛ p ∨ q ∧ r
+  exact test_sorry
+
+-- `exists_or` and `forall_and` cannot be used by `pull` when `∃`/`∀` is only on one side
```
doesn't that just mean that we should mark `forall_and_left` and `forall_and_right` with `@[push]`?

JovanGerb replied:
Good point, I have now marked them with `@[push low]`, so that they are tried after `forall_and` by the `pull` tactic. And I've updated the tests so that it breaks without that `low` priority.

JovanGerb committed:
`root` -> `head`

JovanGerb committed:
avoid `Array.back!`

JovanGerb committed:
Merge branch 'master' into push_tactic

JovanGerb committed:
feat: tag `forall_and_left` and `forall_and_right`

adomani commented on Mathlib/Tactic/Push/Attr.lean:
````diff
@@ -0,0 +1,120 @@
+  }
+
+/--
+The `push` attribute is used to tag lemmas that "push" a constant into an expression.
+
+For example:
+```lean
+@[push] theorem not_not (p : Prop) : ¬ ¬p ↔ p
+
+@[push] theorem not_imp (p q : Prop) : ¬(p → q) ↔ p ∧ ¬q
+
+@[push] theorem not_iff (p q : Prop) : ¬(p ↔ q) ↔ (p ∧ ¬q) ∨ (¬p ∧ q)
+```
+
````
````suggestion
The `push` attribute is used to tag lemmas that "push" a constant into an expression (or makes it disappear).

For example:
```lean
@[push] theorem not_imp (p q : Prop) : ¬(p → q) ↔ p ∧ ¬q

@[push] theorem not_iff (p q : Prop) : ¬(p ↔ q) ↔ (p ∧ ¬q) ∨ (¬p ∧ q)

@[push] theorem not_not (p : Prop) : ¬ ¬p ↔ p
```
In the `not_not` example, the two `not`s cancelled each other, and we consider this as having `push`ed them out of the expression
````
Or maybe some variation?  Specifically, I would rather have the `not_not` example be last and also have an explanation of why this counts as `push`ing.

When the name was `push_neg`, it was kind of clear what was going on, but with a "generic" `push`, I initially found this confusing.

JovanGerb replied:
Thanks, I've updated the doc-string

JovanGerb committed:
update `@[push]` doc-string

JovanGerb committed:
typo

JovanGerb committed:
Merge branch 'master' into push_tactic

adomani commented on Mathlib/Tactic/Push.lean:
```diff
@@ -1,216 +1,348 @@
+macro (name := pushNegCommand) tk:"#push_neg " e:term : command => `(command| #push%$tk Not $e)
+
+@[inherit_doc pull]
+elab "pull " disch?:(discharger)? head:(colGt term) : conv => withMainContext do
```
```suggestion
elab "pull " disch?:(discharger ppSpace)? head:(colGt term) : conv => withMainContext do
```

adomani commented on Mathlib/Tactic/Push.lean:
```diff
@@ -1,216 +1,348 @@
-using say `push_neg at h h' ⊢` as usual.
+`#pull` understands local variables, so you can use them to introduce parameters.
+-/
+macro (name := pullCommand) tk:"#pull " head:ident e:term : command =>
```
```suggestion
macro (name := pullCommand) tk:"#pull " head:ident ppSpace e:term : command =>
```

adomani reviewed (commented):
Sprinkling some `ppSpace`s.

JovanGerb committed:
add missing `ppSpace`

JovanGerb commented:
Thanks for the reviews!

For the `Polynomial.coeff` example, it should definitelly be possible to turn `coeff (X ^ 2 + 3 * X) 1` into something like `(if 1 = 2 then 1 else 0) + 3 * 1`. The issue is that you will need to do some simplification afterwards, so I think you might be better off just using `simp`.

JovanGerb committed:
.

JovanGerb commented on Mathlib/Data/Set/Basic.lean:
```diff
@@ -485,21 +485,35 @@ instance uniqueEmpty [IsEmpty α] : Unique (Set α) where
+theorem not_nonempty_iff_eq_empty : ¬s.Nonempty ↔ s = ∅ := by
   simp only [Set.Nonempty, not_exists, eq_empty_iff_forall_notMem]
 
 /-- See also `Set.not_nonempty_iff_eq_empty`. -/
```
I would rather just remove these doc-string. I think they are redundant.

adomani replied:
Ok, let's leave this as is.  In general, rather than removing doc-strings, I would prefer to improve them, but leaving them unchanged is always an easy option!

JovanGerb committed:
remove wrong `@[push]` annotations

JovanGerb committed:
use commands in test file

JovanGerb commented:
I didn't put the command `#push_discr_tree Not` into the test file, because I think this will be annoying to update when tagging more `push_neg` lemmas. However, looking at the output made me realize that some `@[push]` tags that I added weren't supposed to be there, so I've removed them now.

JovanGerb commented:
For a tactic that computes `Polynomial.coeff`, if you want to restrict your `simp` to a `simp only`, then you could indeed define a `pushCoeff` simproc to be an argument to the `simp only`. We can make a `pushFoo` simproc for any constant `Foo` where that is deemed useful, similar to `pushNeg` and `pushFun`.

adomani commented on Mathlib/Tactic/Push.lean:
```diff
@@ -1,216 +1,350 @@
+  transformAtLocation (pullCore head · disch?) "pull" loc (failIfUnchanged := true) false
+
+/-- A simproc variant of `push fun _ ↦ _`, to be used as `simp [↓pushFun]`. -/
+simproc_decl _root_.pushFun (fun _ ↦ ?_) := pushStep .lambda
```
Could you also add tests for the two `simproc`s, please?  :upside_down_face:

adomani reviewed (commented):
I like the new tests, thanks!  I also have a couple more that I just noticed...

I do not have a feeling for how much the `#push_discr_tree` test will cause issues.  If it turns out to be quite fussy, there is always the option of commenting it out, while we think of an alternative.

Thank you for your patience through the various reviews!

bors d+

JovanGerb committed:
minor refactor

JovanGerb commented:
bors r+

JovanGerb committed:
Merge branch 'master' into push_tactic
